pyomo/solvers/plugins/solvers/mosek_direct.py

```diff
@@ -137,316 +149,288 @@ def _process_stream(msg):
             msg = e.args[0]
-            msg += '\nexpr: {0}'.format(expr)
-            raise DegreeError(msg)
-
-        return mosek_expr, referenced_vars
+            msg += '\nexpr: {}'.format(expr)
+            logger.error(DegreeError(msg))
+            raise e
+        return mosek_arow, mosek_qexp, referenced_vars
+
+    def _mosek_vartype_from_var(self, var):
+        if var.is_integer():
+            return self._mosek.variabletype.type_int
+        return self._mosek.variabletype.type_cont
+
+    def _mosek_bounds(self, lb, ub, fixed_bool):
+        if fixed_bool:
+            return self._mosek.boundkey.fx
+        if lb == -inf:
+            if ub == inf:
+                return self._mosek.boundkey.fr
+            else:
+                return self._mosek.boundkey.up
+        elif ub == inf:
+            return self._mosek.boundkey.lo
+        return self._mosek.boundkey.ra
 
     def _add_var(self, var):
-        varname = self._symbol_map.getSymbol(var, self._labeler)
-        vtype = self._mosek_vtype_from_var(var)
-        if var.has_lb():
-            lb = value(var.lb)
-        else:
-            lb = '0'
-        if var.has_ub():
-            ub = value(var.ub)
-        else:
-            ub = '0'
-
-        bound_type = self.set_var_boundtype(var, ub, lb)
-        self._solver_model.appendvars(1)
-        index = self._solver_model.getnumvar()-1
-        self._solver_model.putvarbound(index, bound_type, float(lb), float(ub))
-        self._solver_model.putvartype(index, vtype)
-        self._solver_model.putvarname(index, varname)
+        """
+        Prepare and pass one variable to MOSEK.
 
-        self._pyomo_var_to_solver_var_map[var] = index
-        self._solver_var_to_pyomo_var_map[index] = var
-        self._referenced_variables[var] = 0
+        This will keep any existing model components intact.
 
-    def _set_instance(self, model, kwds={}):
-        self._range_constraints = set()
-        DirectOrPersistentSolver._set_instance(self, model, kwds)
-        self._pyomo_con_to_solver_con_map = dict()
-        self._solver_con_to_pyomo_con_map = ComponentMap()
-        self._pyomo_cone_to_solver_cone_map = dict()
-        self._solver_cone_to_pyomo_cone_map = ComponentMap()
-        self._pyomo_var_to_solver_var_map = ComponentMap()
-        self._solver_var_to_pyomo_var_map = ComponentMap()
-        self._whichsol = getattr(
-            self._mosek.soltype, kwds.pop('soltype', 'bas'))
+        Parameters
+        ----------
+        var: Var
+        """
+        self.add_vars((var,))
 
-        try:
-            self._solver_model = self._mosek_env.Task(0, 0)
-        except Exception:
-            e = sys.exc_info()[1]
-            msg = ("Unable to create Mosek Task. "
-                   "Have you installed the Python "
-                   "bindings for Mosek?\n\n\t" +
-                   "Error message: {0}".format(e))
-            raise Exception(msg)
+    def add_vars(self, var_seq):
```

Comment:
Agreed, I will make the change. 
However, I have a (sort-of) unrelated question about the wrapper method for adding constraints: if a user decides to add a cone, but one that is constructed as a domain, then they would be essentially adding a block. Do you think, then, that it would be a good idea to raise an error telling the user to switch to the add_block method, or would it be better to just call the add_block method from within the add_constraints wrapper method?

Reply:
That is an excellent question. I would lean toward calling add_block within the add_constraints wrapper, but I don't have a strong preference.